runners/apex/src/main/java/org/apache/beam/runners/apex/ApexYarnLauncher.java

```diff
@@ -0,0 +1,383 @@
+        char c = line.charAt(i);
+        if (Character.isLetter(c)) {
+          if (i > excludeLevel) {
+            excludes.add(line.substring(i));
+          } else {
+            if (line.substring(i).startsWith("org.apache.hadoop")) {
+              excludeLevel = i;
+              excludes.add(line.substring(i));
+            } else {
+              excludeLevel = Integer.MAX_VALUE;
+            }
+          }
+          break;
+        }
+      }
+    }
+    br.close();
+
+    Set<String> excludeJarFileNames = Sets.newHashSet();
+    for (String exclude : excludes) {
+      String[] mvnc = exclude.split(":");
+      String fileName = mvnc[1] + "-";
+      if (mvnc.length == 6) {
+        fileName += mvnc[4] + "-" + mvnc[3]; // with classifier
+      } else {
+        fileName += mvnc[3];
+      }
+      fileName += ".jar";
+      excludeJarFileNames.add(fileName);
+    }
+
+    ClassLoader classLoader = ApexYarnLauncher.class.getClassLoader();
+    URL[] urls = ((URLClassLoader) classLoader).getURLs();
+    List<File> dependencyJars = new ArrayList<>();
+    for (int i = 0; i < urls.length; i++) {
+      File f = new File(urls[i].getFile());
+      // dependencies can also be directories in the build reactor,
+      // the Apex client will automatically create jar files for those.
+      if (f.exists() && !excludeJarFileNames.contains(f.getName())) {
+          dependencyJars.add(f);
+      }
+    }
+    return dependencyJars;
+  }
+
+  /**
+   * Create a jar file from the given directory.
+   * @param dir source directory
+   * @param jarFile jar file name
+   * @throws IOException when file cannot be created
+   */
+  public static void createJar(File dir, File jarFile) throws IOException {
+
+    final Map<String, ?> env = Collections.singletonMap("create", "true");
+    if (!jarFile.delete()) {
+      throw new RuntimeException("Failed to remove " + jarFile);
+    }
+    URI uri = URI.create("jar:" + jarFile.toURI());
+    try (final FileSystem zipfs = FileSystems.newFileSystem(uri, env);) {
+      final java.nio.file.Path root = dir.toPath();
+      Files.walkFileTree(root, new java.nio.file.SimpleFileVisitor<Path>() {
+        String relativePath;
+
+        @Override
+        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+            throws IOException {
+          relativePath = root.relativize(dir).toString();
+          if (!relativePath.isEmpty()) {
+            if (!relativePath.endsWith("/")) {
+              relativePath += "/";
+            }
+            final Path dstDir = zipfs.getPath(relativePath);
+            Files.createDirectory(dstDir);
+          }
+          return super.preVisitDirectory(dir, attrs);
+        }
+
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+          String name = relativePath + file.getFileName();
+          if (!JarFile.MANIFEST_NAME.equals(name)) {
```

Comment:
This excludes MANIFEST as it should be the first entry in the .jar file, but the code to add MANIFEST as the first entry is missing. Is this intentional?